src/rtos/task_notify_when_deleting.c

```diff
@@ -41,15 +44,39 @@ void task_notify_when_deleting(task_t target_task, task_t task_to_notify,
   task_to_notify = (task_to_notify == NULL) ? pxCurrentTCB : task_to_notify;
   target_task = (target_task == NULL) ? pxCurrentTCB : target_task;
 
-  // It doesn't make sense for a task to notify itself, and make sure that neither task is NULL (implying that scheduler hasn't started yet)
+  // It doesn't make sense for a task to notify itself, and make sure that
+  // neither task is NULL (implying that scheduler hasn't started yet)
   if (task_to_notify == target_task || !task_to_notify || !target_task) {
     return;
   }
 
-  linked_list_s_t* target_ll = pvTaskGetThreadLocalStoragePointer(target_task, TLSP_IDX);
+  // task_to_notify maintains a list of the tasks whose deletion it cares about.
+  // This will allow us to unsubscribe from notification if/when task_to_notify
+  // is deleted
+  linked_list_s_t* subscriptions_ll = pvTaskGetThreadLocalStoragePointer(task_to_notify, SUBSCRIPTIONS_TLSP_IDX);
+  if (subscriptions_ll == NULL) {
+    subscriptions_ll = linked_list_init();
+    vTaskSetThreadLocalStoragePointer(task_to_notify, SUBSCRIPTIONS_TLSP_IDX, subscriptions_ll);
+  }
+  if (subscriptions_ll != NULL) {
+    // check whether task_to_notify is already subscribed. if so, do nothing
+    ll_node_s_t* it = subscriptions_ll->head;
+    bool found = false;
+    while (it != NULL && !found) {

```

Comment:
And we search through that linked list for `task_to_notify`? Why?

Reply:
```c
// check whether task_to_notify is already subscribed. if so, do nothing
```

Reply:
yeah oops. too many conflicting terminologies